image/cas/layout/tar.go

```diff
@@ -0,0 +1,108 @@
+// Copyright 2016 The Linux Foundation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package layout
+
+import (
+	"bytes"
+	"crypto/sha256"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"strings"
+
+	"github.com/opencontainers/image-spec/image/cas"
+	"github.com/opencontainers/image-spec/image/layout"
+	"golang.org/x/net/context"
+)
+
+// TarEngine is a cas.Engine backed by a tar file.
+type TarEngine struct {
+	file ReadWriteSeekCloser
+}
+
+// NewTarEngine returns a new TarEngine.
+func NewTarEngine(ctx context.Context, file ReadWriteSeekCloser) (eng cas.Engine, err error) {
+	engine := &TarEngine{
+		file: file,
+	}
+
+	err = layout.CheckTarVersion(ctx, engine.file)
+	if err != nil {
+		return nil, err
+	}
+
+	return engine, nil
+}
+
+// Put adds a new blob to the store.
+func (engine *TarEngine) Put(ctx context.Context, reader io.Reader) (digest string, err error) {
+	data, err := ioutil.ReadAll(reader)
```

Comment:
Let's avoid allocating all the memory for this.

The idea would be to string the read of this content through the hash calculation, into a temp file.